tests: thrift: tlsTransport:

**Describe the bug**

The Thrift `tlsTransport` test is failing in CI for `qemu_x86_64` and possibly other platforms

https://github.com/zephyrproject-rtos/zephyr/actions/runs/5697038294/job/15443170454?pr=60942#step:13:282

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? `qemu_x86_64`
 - What have you tried to diagnose or workaround this issue? manual stress tests
[run-thrift-trials.sh.txt](https://github.com/zephyrproject-rtos/zephyr/files/12201792/run-thrift-trials.sh.txt)
 - Is this a regression? If yes, have you been able to "git bisect" it to a
   specific commit? Unknown at this time
 - ...

**To Reproduce**
Steps to reproduce the behavior:
* `twister -i -p qemu_x86_64 -s tests/lib/thrift/ThriftTest`
* `./run-thrift-trials.sh`

**Expected behavior**
It should work to a reasonable degree of 9's.

**Impact**
It was holding up CI today

**Logs and console output**
In this test, there is some 
```
INFO    - 221/617 qemu_x86_64               tests/lib/thrift/ThriftTest/thrift.ThriftTest.newlib.tlsTransport  FAILED unexpected eof (qemu 6.919s)
...
PASS - test_map in 0.518 seconds
===================================================================
START - test_multi_exception
Thrift: Thu Jan  1 00:00:06 1970 TServerSocket::acceptImpl() ::accept() : Software caused connection abort

Thrift: Thu Jan  1 00:00:06 1970 TSocket::open() connect() <Host: 192.0.2.1 Port: 4242>: Software caused connection abort

abort()
E: RAX: 0x0000000000000004 RBX: 0x00000000006fcbe0 RCX: 0x0000000000000001 RDX: 0x0000000000000000
E: RSI: 0x00000000000003f8 RDI: 0x00000Thrift: 000001febf0 RBP: 0x0000000000206cf0 RSP: 0xThu Jan  1 00:00:06 1970 TServerTransport died: accept(): Software caused connection abort

0000000000206cf0
E:  R8: 0x0000000000000000  R9: 0x00000000001c59fe R10: 0x0000000000206630 R11: 0x000000000000000b
E: R12: 0x00000000006fca60 R13: 0x0000000000206dd0 R14: 0x00000000001ed920 R15: 0x0000000000000000
E: RSP: 0x0000000000206cf0 RFLAGS: 0x0000000000000203 CS: 0x0018 CR3: 0x0000000000232000
E: RIP: 0x0000000000177ea7
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x1ec[400](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5697038294/job/15443170454?pr=60942#step:13:401) (test_multi_exception)
E: Halting system
```

In the test below, things seem to just hang.
```
617/617 qemu_x86_64               tests/lib/thrift/ThriftTest/thrift.ThriftTest.newlib.tlsTransport  FAILED Timeout (qemu 59.936s)
...
testByte(-128)

 PASS - test_byte in 0.129 seconds
===================================================================
START - test_double
```

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows) Linux
 - Toolchain (e.g Zephyr SDK, ...): Zephyr SDK 0.16.1
 - Commit SHA or Version used: 488fd89e2bd64c60354378fdec33b7e65c713d38

**Additional context**

This is semi-related to https://github.com/zephyrproject-rtos/zephyr/issues/60941 and https://github.com/zephyrproject-rtos/zephyr/issues/60895

With this patch
```diff
diff --git a/lib/os/fdtable.c b/lib/os/fdtable.c
index 1d5c340cc1..639e54511f 100644
--- a/lib/os/fdtable.c
+++ b/lib/os/fdtable.c
@@ -87,6 +87,7 @@ static int z_fd_unref(int fd)
                return old_rc - 1;
        }
 
+       printk("FD: FREE: %d\n", fd);
        fdtable[fd].obj = NULL;
        fdtable[fd].vtable = NULL;
 
@@ -213,6 +214,8 @@ int z_reserve_fd(void)
                fdtable[fd].vtable = NULL;
                k_mutex_init(&fdtable[fd].lock);
                k_condvar_init(&fdtable[fd].cond);
+
+               printk("FD: RSRV: %d\n", fd);
        }
 
        k_mutex_unlock(&fdtable_lock);
```

On `main`, from running the stress test, we see that eventually there is a failure, and the first file descriptor that is allocated in this case is 11. Normally, this should be 3 (0, 1, and 2 are stdin, stdout, and stderr, respectively).
```
********************************************************************************
Job 6/100                                       [FAIL] Pass Rate: 85 %
FD: RSRV: 11
FD: RSRV: 12
FD: RSRV: 13
FD: RSRV: 14
testMultiException(Xception, test 1)
testMultiException(Xception2, test 2)
FD: FREE: 14
FD: FREE: 13
FD: FREE: 8
FD: FREE: 12
FD: FREE: 11
FD: FREE: 5
FD: FREE: 6
FD: FREE: 7
FD: FREE: 10
FD: FREE: 9
FD: FREE: 4
FD: FREE: 3
 PASS - test_multi_exception in 0.043 seconds
===================================================================
START - test_nested_map
FD: RSRV: 3
FD: RSRV: 4
FD: RSRV: 5
FD: RSRV: 6
FD: RSRV: 7
FD: RSRV: 8
FD: RSRV: 9
FD: RSRV: 10
FD: RSRV: 11
FD: RSRV: 12
FD: RSRV: 13
FD: RSRV: 14
testMapMap(1)
FD: FREE: 14
FD: FREE: 13
FD: FREE: 8
FD: FREE: 12
FD: FREE: 11
FD: FREE: 5
FD: FREE: 6
FD: FREE: 7
FD: FREE: 10
FD: FREE: 9
FD: FREE: 4
FD: FREE: 3
 PASS - test_nested_map in 0.042 seconds
===================================================================
START - test_nested_struct
FD: RSRV: 3
FD: RSRV: 4
FD: RSRV: 5
FD: RSRV: 6
FD: RSRV: 7
FD: RSRV: 8
FD: RSRV: 9
FD: RSRV: 10
FD: RSRV: 11
FD: RSRV: 12
FD: RSRV: 13
FD: RSRV: 14
ninja: build stopped: interrupted by user.
qemu-system-x86_64: terminating on signal 15 from pid 3304336 (timeout)
```

After the fix in #60956, we see that all test cases begin reserving file descriptors at 3 again.
[log.txt](https://github.com/zephyrproject-rtos/zephyr/files/12209028/log.txt)

--- a/tests/lib/thrift/ThriftTest/overlay-tls.conf
+++ b/tests/lib/thrift/ThriftTest/overlay-tls.conf
@@ -1,5 +1,21 @@
 CONFIG_THRIFT_SSL_SOCKET=y
 
+# Currenty, in Zephyr's MBedTLS IPPROTO_TLS_1_0 implementation, 2 sockets are
+# needed for every connection.
+#
+# Additionally, upstream Apache Thrift uses socketpair for cancellation rather
+# than eventfd, since the latter is not portable to some operating systems.
+#
+# File Descriptor Usage
+# ---------------------
+# stdin, stdout, stderr: 3
+# tcp socket (accept): 1
+# tls socket (accept): 1
+# tcp sockets (client, server): 2
+# tls sockets (client, server): 2
+# socketpairs for cancellation (accept, client, server): 6
+CONFIG_POSIX_MAX_FDS=15
+
 # TLS configuration
 CONFIG_MBEDTLS=y
 CONFIG_MBEDTLS_PEM_CERTIFICATE_FORMAT=y

--- a/tests/lib/thrift/ThriftTest/prj.conf
+++ b/tests/lib/thrift/ThriftTest/prj.conf
@@ -39,7 +39,19 @@ CONFIG_NET_BUF_TX_COUNT=20
 CONFIG_NET_PKT_TX_COUNT=20
 CONFIG_NET_BUF_RX_COUNT=20
 CONFIG_NET_PKT_RX_COUNT=20
-CONFIG_POSIX_MAX_FDS=16
+
+# We can get away with using fewer sockets in the non-TLS tests because we use
+# TFDServer.cpp for our server and socketpair() for our channel. We do not
+# need an accept socket for the server (in contrast to TCP), it only needs 1
+# eventfd for server cancellation, and there are no cancellation sockets
+# required because we close them in the testsuite.
+#
+# File Descriptor Usage
+# ---------------------
+# stdin, stdout, stderr: 3
+# socketpair for channel: 2
+# eventfd for cancellation: 1
+CONFIG_POSIX_MAX_FDS=6
 
 # Network address config
 CONFIG_NET_IPV4=y

--- a/tests/lib/thrift/ThriftTest/src/main.cpp
+++ b/tests/lib/thrift/ThriftTest/src/main.cpp
@@ -153,13 +153,13 @@ static void thrift_test_after(void *data)
 
 	pthread_join(context.server_thread, NULL);
 
-	context.server.reset();
-	context.client.reset();
-
 	for (auto &fd : context.fds) {
 		close(fd);
 		fd = -1;
 	}
+
+	context.client.reset();
+	context.server.reset();
 }
 
 ZTEST_SUITE(thrift, NULL, thrift_test_setup, thrift_test_before, thrift_test_after, NULL);

